### PR TITLE
[MONO] Implement GC.AddMemoryPressure in Mono

### PR DESCRIPTION
**Pull request by naricc**
This implements [GC.AddMemoryPressure](https://learn.microsoft.com/en-us/dotnet/api/system.gc.addmemorypressure?view=net-7.0) in Mono. It basically just copies the heuristics around this API from the CoreCLR implementation and maps it to sgen: https://github.com/dotnet/runtime/blob/f857468fd9ae571ca78ee0f22915e2489404541a/src/coreclr/vm/comutilnative.cpp#L1261.

Those heuristics may need tuning for mono work loads, but we need some implementation to even begin that work. This seems like a good starting point.

This contributes to missing the missing GC API issue: https://github.com/dotnet/runtime/issues/73167

### CODE REVIEW

**Comment by naricc** [comment]
@Redth @jonpryor Here is the implementation of AddMemoryPressure in Mono that we discussed briefly in the Cross Team Sync. The CoreCLR version had a bunch of heuristics around it, and I basically just copied those; I suspect we will have to update them as we get a better idea of use cases.

File: src/mono/mono/metadata/gc-internals.h [hunk 1]
```diff
@@ -170,6 +170,8 @@ typedef void (*MonoFinalizationProc)(gpointer, gpointer); // same as SGenFinaliz
 
 void  mono_gc_register_for_finalization (MonoObject *obj, MonoFinalizationProc user_data);
 void  mono_gc_add_memory_pressure (gint64 value);
+void  mono_gc_remove_memory_pressure (gint64 value);
```
Comment (AaronRobinsonMSFT) [inline src/mono/mono/metadata/gc-internals.h:173]:
```suggestion
void  mono_gc_add_memory_pressure (guint64 value);
void  mono_gc_remove_memory_pressure (guint64 value);
```

Should be `guint64` in the impl too. Following the managed signature with `ulong`.

File: src/mono/mono/sgen/sgen-gc.c [hunk 2]
```diff
@@ -4083,4 +4084,96 @@ sgen_check_canary_for_object (gpointer addr)
 	}
 }
 
+
+const guint MEM_PRESSURE_COUNT = 4;
```
Comment (BrzVlad) [inline src/mono/mono/sgen/sgen-gc.c:4088]: all the code around here doesn't use mono code style

**Comment by BrzVlad** [comment]
Isn't adding memory pressure the same as if we had an object allocated with that size. And have the same GC heuristic as if there is more memory allocated ? Although in that case, in the long term it would lead for fewer GCs on mono.

**Comment by naricc** [comment]
> Isn't adding memory pressure the same as if we had an object allocated with that size. And have the same GC heuristic as if there is more memory allocated ? Although in that case, in the long term it would lead for fewer GCs on mono.

I expected it to work that way too. But after looking at the CoreCLR implementation, and talking to @AaronRobinsonMSFT, the existing implementation turns out not to do it that way. Its basically a heuristic around GC.Collect, with some information from the user. If the user keeps AddingMemoryPressure, it may force a collection, if it would not add too much GC overhead. But it doesn't actually change the GCs normal calculation.


Of course, I could make the mono implementation different. But I elected to follow the existing implementation to start with.

File: src/mono/mono/sgen/sgen-gc.c (hunk 2 above)
Comment (naricc) [reply src/mono/mono/sgen/sgen-gc.c]: I think I got them all.

File: src/mono/mono/sgen/sgen-gc.c [hunk 3]
```diff
@@ -4083,4 +4084,96 @@ sgen_check_canary_for_object (gpointer addr)
...
+
+    guint p = memory_pressure_iteration % MEM_PRESSURE_COUNT;
+
+    guint64 new_mem_value = mono_atomic_fetch_add_i64 ((gint64*)&memory_pressure_adds[p], bytes_allocated);
+    
+    // AddMemoryPressure contains unrolled loops which depend on MEM_PRESSURE_COUNT
+    g_assert (MEM_PRESSURE_COUNT == 4);
+
+    guint64 add = memory_pressure_adds[0] + memory_pressure_adds[1] + memory_pressure_adds[2] + memory_pressure_adds[3] - memory_pressure_adds[p];
+    guint64 rem = memory_pressure_removes[0] + memory_pressure_removes[1] + memory_pressure_removes[2] + memory_pressure_removes[3] - memory_pressure_removes[p];
```
Comment (lambdageek) [inline src/mono/mono/sgen/sgen-gc.c:4133]: This is silly. The compiler can unroll a loop. It's just needlessly fragile. (For mono. I'm sure coreclr measured and found this to be a problem on some hot path)

File: src/mono/mono/sgen/sgen-gc.c [hunk 4]
```diff
@@ -4083,4 +4084,96 @@ sgen_check_canary_for_object (gpointer addr)
...
+guint64 memory_pressure_removes[MEM_PRESSURE_COUNT] = {0, 0, 0, 0};   // history of memory pressure removals
+const unsigned min_memorypressure_budget = 4 * 1024 * 1024;        // 4 MB
+
+// Resets pressure accounting after a gen2 GC has occurred.
+static void check_pressure_counts ()
+{
+    if (memory_pressure_gc_count != sgen_gc_collection_count(GENERATION_OLD))
+    {
+        memory_pressure_gc_count = sgen_gc_collection_count(GENERATION_OLD);
+        memory_pressure_iteration++;
```
Comment (lambdageek) [inline src/mono/mono/sgen/sgen-gc.c:4102]: This should probably be an atomic increment

File: src/mono/mono/sgen/sgen-gc.c [hunk 5]
```diff
@@ -4083,4 +4084,96 @@ sgen_check_canary_for_object (gpointer addr)
...
+
+    guint64 new_mem_value = mono_atomic_fetch_add_i64 ((gint64*)&memory_pressure_adds[p], bytes_allocated);
+    
+    // AddMemoryPressure contains unrolled loops which depend on MEM_PRESSURE_COUNT
+    g_assert (MEM_PRESSURE_COUNT == 4);
+
+    guint64 add = memory_pressure_adds[0] + memory_pressure_adds[1] + memory_pressure_adds[2] + memory_pressure_adds[3] - memory_pressure_adds[p];
+    guint64 rem = memory_pressure_removes[0] + memory_pressure_removes[1] + memory_pressure_removes[2] + memory_pressure_removes[3] - memory_pressure_removes[p];
+
+    if (new_mem_value >= min_memorypressure_budget)
```
Comment (lambdageek) [inline src/mono/mono/sgen/sgen-gc.c:4135]: The rest of the code here seems to do two things

1. Evaluates a heuristic to decide if it should run a collection
2. Runs a collection.

Can we make those into separate `static` functions

File: src/mono/mono/sgen/sgen-gc.c (hunk 3 above)
Comment (naricc) [reply src/mono/mono/sgen/sgen-gc.c:4133]: Yeah, I guess I copied a little too verbatim. Will re-roll.

File: src/mono/mono/sgen/sgen-gc.c [hunk 6]
```diff
@@ -4091,89 +4090,92 @@ guint64 memory_pressure_gc_count = 0;
...
-                {
-		    sgen_gc_collect (GENERATION_OLD);
-                    check_pressure_counts ();
+	guint64 add = 0;
+	guint64 rem = 0;
+
+	for (gint i = 0; i < MEM_PRESSURE_COUNT; i++) {
+		add += memory_pressure_adds[i];
+		rem += memory_pressure_removes[i];
+	}
```
Comment (lambdageek) [inline src/mono/mono/sgen/sgen-gc.c:4128]:
```c
  add -= memory_pressure_adds[p];
  rem -= memory_pressure_removes[p];
```

**Comment by lambdageek** [comment]
We talked this morning about going another way, and pulling this chunk of `GC.NativeAot.cs` into `src/libraries/.../GC.MemoryPressure.cs` and sharing it between NativeAOT and Mono.  There's one bit (a couple of calls to `RuntimeImports.RhXYZ` to decide if we're really, finally, going to trigger a GC) that will need to be runtime-specific, but the rest can be shared

https://github.com/dotnet/runtime/blob/56a271e2b7743ddeefc2329ff1435832ca4efd7a/src/coreclr/nativeaot/System.Private.CoreLib/src/System/GC.NativeAot.cs#L421-L553

**Review by lambdageek — APPROVED** [review]

File: src/mono/mono/sgen/sgen-gc.c (hunk 3 above)
Comment (AaronRobinsonMSFT) [reply src/mono/mono/sgen/sgen-gc.c]: > I'm sure coreclr measured and found this to be a problem on some hot path

That is very generous of you. It is quite possible, but if it was it more than a decade ago. I too am skeptical it has any value nowadays.

**Pull request merged**